### DISCUSSION TITLE
Avoiding timeouts in long running queries

### DISCUSSION THREAD

**Discussion by TheOtherBrian1** [supabase/supabase#21133] [Troubleshooting]
_This is a copy of a troubleshooting article on Supabase's docs site. It may be missing some details from the original. View the [original article](https://supabase.com/docs/guides/troubleshooting/avoiding-timeouts-in-long-running-queries-6nmbdN)._

Certain queries, like indexing a table or changing a column's data type, are inherently time-consuming. The [Dashboard's SQL Editor](https://supabase.com/dashboard/project/_/sql/) has a 1-minute timeout limit.

To execute long-running queries, follow the below steps.

## Install an external SQL client

The guide focuses on [psql](https://supabase.com/docs/guides/database/psql) but you can use any Postgres client.

Some other options include:

*   [pgAdmin](https://supabase.com/docs/guides/database/pgadmin)
*   [DBeaver](https://supabase.com/docs/guides/database/dbeaver)

You can install PSQL in [macOS](https://stackoverflow.com/a/49689589/2188186) and [Windows](https://www.postgresql.org/download/windows/) by following these links and instructions. For Linux (Debian) you can run the following:

```bash
sudo apt-get update
sudo apt-get install postgresql-client
```

Once installed, you can find your PSQL string on the dashboard by clicking [connect](https://supabase.com/dashboard/project/_?showConnect=true). Make sure if you are using the pooler connection that it is the [Session pooler](https://supabase.com/dashboard/project/_?showConnect=true\&method=session) (port 5432).

## Increase the query timeout

Then you can increase the query timeout solely for your session:

Long timeout

```sql
set statement_timeout = '120min';
```

Disable timeout

```sql
set statement_timeout = '0';
```

If your task is particularly long, you can may want to consider boosting your computing power temporarily. Compute size on Supabase is charged by the hour, so you can increase it for an hour or two, complete your task faster, then scale it back afterwards.

If you want to temporarily upgrade, you can find the add-ons for your project in your [Dashboard's Add-ons Settings.](https://supabase.green/dashboard/project/_/settings/addons)

**Comment by grieblius** [comment]
Hi,
I have several long running routine functions which are executed from external source via Supabase JS API (PostgREST).
Now I'm stuck with 8s [statement_timeout](https://postgresqlco.nf/doc/en/param/statement_timeout/).

Is there any possibility to set statement_timeout for some specific function? I've tried to do `set local statement_timeout = '2min';` inside function, but it didn't work.

I understand that there is possibility to set different timeouts for each role and I can set timeout for service role, but sometimes service role can be used for specific queries to provide data for public users and I would want to allow bigger timeout only for specific tasks.

**Reply by TheOtherBrian1** [reply]
I'm not sure if it is possible to configure the timeout per transaction, but it is a good question. If you want, I can ask internally, but it may be better to ask this question in the [postgrest-js](https://github.com/supabase/postgrest-js) or [postgREST](https://github.com/PostgREST/postgrest/discussions) repo

**Reply by grieblius** [reply]
Thank you,

seems found exactly what I need - it is possible to apply `statement_timeout` for function:

```sql
CREATE OR REPLACE FUNCTION myfunc()
RETURNS void as $$
  SELECT pg_sleep(3); -- simulating some long-running process
$$
LANGUAGE SQL
SET statement_timeout TO '4s';
```

Found this in [PostgREST docs](https://postgrest.org/en/latest/references/transactions.html#function-settings).

**Comment by MCYouks** [comment]
Thanks for sharing!
In the solution provided, since the script is run from the SQL client, the index creation won't appear in the list of migration files. How would you handle it?

Thank you!

**Reply by TheOtherBrian1** [reply]
Hi @MCYouks, 

You can set the session timeout in your migration script:

```sql
SET statement_timeout = '20min'; --<----modify time as needed
```

Then, as long as you're using the direct connection or Supavisor in session mode (both use port 5432), you should be able to run the migration as normal.
